try to fix the issue #202 and #220:

nanli-emory opened the pull request:
1. wrap 'getBestThumb' to handle whether the slide has a bounding box or not
2. wrap 'getBestLevelForDownsample' to handle numerical precision for getting an appropriate level
3. wrap 'rgba2rgb' to handle merge alpha channel
4. add 'setBBox' to get the bounding box info (x, y, width, height). If there is no bounding box info, then set the bounding box as (0,0, dimensions[0], dimensions[1])
5. add 'img_bbox' property to save the bounding box info

nanli-emory committed:
try to fix the issue #202 and #220:
1. wrap 'getBestThumb' to handle that the slide has bounding box or not
1. wrap 'getBestLevelForDownsample' to handle numerical precision for getting an appropriate level
2. wrap 'rgba2rgb' to handle merge alpha channel
3. add 'setBBox' to get the the bounding box info (x, y, width, height) if there is no bounding box info then set the bounding box as (0,0, dimensions[0], dimensions[1])
4. add 'img_bbox' property to save the bounding box info

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -1,13 +1,28 @@
 import openslide
+def getBestThumb(s, x, y, size):
+    osh = s["os_handle"]
+    if s["bounding_box"]:
```
a small point here,  if the "else" code is quite small ( in this case 1 line) and includes a return statement, its better programming style to say:

```
if NOT condition
      return

-rest- of code
```

which eliminates the "else" statement entirely doesn't have a "dangling return" at the end of the function

it also reduces potential indentation errors and makes code readability/undrestanding a bit easier

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -61,7 +76,10 @@ def __init__(self, fname, fname_outdir, params):
         self["dir"] = os.path.dirname(fname)
 
         self["os_handle"] = openslide.OpenSlide(fname)
-        self["image_base_size"] = self["os_handle"].dimensions
```
can we leave this line in for legacy purposes?

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -61,7 +76,10 @@ def __init__(self, fname, fname_outdir, params):
 
         self["os_handle"] = openslide.OpenSlide(fname)
-        self["image_base_size"] = self["os_handle"].dimensions
+        self["bounding_box"] = strtobool(params.get("bounding_box","False"))
```
can you confirm that bounding_box is appearing as a column  in the output TSV file? will definitely want users to have this information in their output

choosehappy replied:
sorry, this confused me, i see now that i was referring to img_bbox appearing in the output file.

can we modify this variable from "bounding_box" to "enable_bounding_box" or something like that, to indicate its flag variable and differentiate it from img_bbox which will contain the actual bounding box?

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -70,7 +88,7 @@ def __init__(self, fname, fname_outdir, params):
         if (self["mask_statistics"] not in mask_statistics_types):
             logging.error(
-                f"mask_statistic type '{self['mask_statistics']}' is not one of the 3 supported options relative2mask, absolute, relative2image!")
+                f"mask_statistic type \"{self['mask_statistics']}\" is not one of the 3 supported options relative2mask, absolute, relative2image!")
```
out of curiosity, did you do this change manually or was it part of an IDE formatting? is there a reason for it or personal preference?

nanli-emory replied:
It is my personal preference. I like to keep consistent **quotation marks** (double or single) in code style/ text output style.

choosehappy replied:
i agree, but looking at the other code i think the single ones are used _in_ a print statement and a double quote is used _around_ the print statement. in this way we don't have to do escape the quotes (which i always fail to do and thus create more errors for myself : )  )

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -90,38 +108,75 @@ def __setitem__(self, key, value):
+            else:
+                # no bbox info in slide set bounding_box as Flase
+                self["bounding_box"] = False
+                logging.warning(f"{self['filename']}: could not read the bounding box")
```
nice, i like this.....maybe we *also* put a warning again the actual image so that we can see this in the tsv file?

something like
self["warnings"].append('Bounding Box requested but could not read')

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -90,38 +108,75 @@ def __setitem__(self, key, value):
+                except:
+                    # no bbox info in slide set bounding_box as Flase
+                    self["bounding_box"] = False
+                    logging.warning(f"{self['filename']}: could not convert the bounding box")
```
similar note as below, but with a different message : )

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -90,38 +108,75 @@ def __setitem__(self, key, value):
+                # no bbox info in slide set bounding_box as Flase
+                self["bounding_box"] = False
+                logging.warning(f"{self['filename']}: could not read the bounding box")
+
```
this is good as it is. in practice however, I'm not sure that there is a meaningful difference on the user side between not being able to convert the bounding box and not being able to read it. if there is no difference, then this would be reduced down to a single cleaner larger try/except without an if

note as well you can use something like this:

xc=int(osh.properties.get(openslide.PROPERTY_NAME_BOUNDS_X,'NA'))

which will definitely cause an error when trying to convert if it doesn't exist (this removes potential 'dangerous' of the default return value of 'get', which some libraries tools could be something like a -1 and you wouldn't detect in your code)

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -1,13 +1,28 @@
+    bg_color = "#" + s["os_handle"].properties.get(openslide.PROPERTY_NAME_BACKGROUND_COLOR, "ffffff")
+    thumb = Image.new("RGB", img.size, bg_color)
+    thumb.paste(img, None, img)
+    return thumb
```
nice! definitely was a bit lazy on my part previously : )

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -131,29 +186,30 @@ def getImgThumb(self, dim):
+                downsample_factor = base_mag / target_mag
+                level = self.getBestLevelForDownsample(downsample_factor)
+                relative_down = downsample_factor/osh.level_downsamples[level]
+                size=tuple((np.array(img_base_size)/osh.level_downsamples[level]).astype(int)) if self["bounding_box"] else img_base_size
```
nice

choosehappy reviewed (changes requested):
Looks great! A few minor points to consider.

btw, how did you test this? did you have multiple WSIs to try it out on?

i have someone working on a project right now that uses histoqc, they can rerun the pipeline with this code modification, and i think...the results should be exactly the same if bbox = False, is that correct?

nanli-emory committed:
fixed the comment
1. change 'bounding_box' -> 'enable_bounding_box'
2. add "image_bounding_box" into results.tsv
3. add the "warnings" message into error.log if the bounding box requested but can't read from slide
4. find a bug on line 197 and change the from  img_base_size -> osh.level_dimensions[level]
5. leave self["image_base_size"] for legacy
